### ISSUE TITLE
Cannot widen type in generics while retaining knowledge of assignability

### ISSUE THREAD

**Issue by UselessPickles** [microsoft/TypeScript#35257]
**Search Terms:**
generics widen type

**Code**

```ts
type Widen<T extends number | string> = T extends number
    ? number
    : T extends string
    ? string
    : T;

class Foo<T extends number | string> {
  public constructor(private readonly validValues: readonly T[]) { }

  // UNEXPECTED ERROR: T is not assignable to Widen<T>
  public isValidValue(value: Widen<T>): value is T {
    return this.validValues.indexOf(value as any) !== -1;
  }
}
```

**Expected behavior:**
The `Widen` type in my example widens number and string types. For any type `T`, `T` is assignable to `Widen<T>`. I expect to be able to rely on this fact in generics.

**Actual behavior:**
The assignability relationship of `T` and `Widen<T>` is lost when `T` is a generic type rather than a concrete type (see `// UNEXPECTED ERROR` comment in example).

**Playground Link:** 
http://www.typescriptlang.org/play/?ssl=5&ssc=7&pln=5&pc=12#code/C4TwDgpgBA6glgEwgOwDwBUoQB7BQgZymQFcBbAIwgCcoAfKA4auZAcwD4oBeKTHPMkLFyVagCgoUqAH4RlGpOkAuPllz4iTFuyVS521mz1RV6ANzjxAYwA2AQwJEAYgHtXGdYOGkFtBobsXADeSmAkFLZw1lDWrsjaJNbArtQAFGAsAG72eFDUEPYI8bYgUDlRCABq9rYkEASqBUUlZegA2gC6AJRQwVAAvlZSAPQjUACqAHIAogAaAAozAMLoMwAiUDMAStsA8ttmUHBEyK7AUI4EcGzI9pHQKbCIKBgcYRFRMSc1lb-1aQq9VU8CQaHQHG6qiB0BOalC0nyEGAJGoyCgwAAFicAHQVRD-Bo41hIbB7ABmgNq9UuRHsyBAvQAhNxeABaACMlikQwGQA

**Comment by UselessPickles** [comment]
If this is "working as intended" or my expectations are otherwise infeasible, is there any other valid way to achieve what I am attempting?

For example, I want `Foo<"A" | "B">.isValidValue()` to accept `string`, but not `number`,  I want  `Foo<1 | 2>.isValidValue()` to accept `number`, but not `string`, and I want  `Foo<"A" | 2>.isValidValue()` to accept `number | string`.

**Comment by jcalz** [comment]
Assignability for unresolved generic conditional types is an existing sore point.  (Does anyone know if there is a canonical issue for this?  This has to be a duplicate or related to something, but I'm not sure.  It's not exactly #26933)  In cases like this I usually hit the compiler over the head with assignability using an intersection:

```ts
  public isValidValue(value: Widen<T>): value is Widen<T> & T {
    return this.validValues.indexOf(value as any) !== -1;
  }
```

**Comment by UselessPickles** [comment]
@jcalz Thanks for the intersection trick! Seems to be working well for me as a work-around (all my dtslint tests passed).

#26933 Looks very related to me.

**Comment by RyanCavanaugh** [comment]
>  For any type `T`, `T` is assignable to `Widen<T>`. 

This is really only possible to know through second-order reasoning about all possible `T` and isn't something TS is capable of constructing proofs for. In fact, it's only really even true because `string & number` is not inhabitable; the same definition of `Widen` for two non-disjoint types does *not* have the same assignability guarantee.

**Comment by UselessPickles** [comment]
@RyanCavanaugh Thanks for the explanation. I see how that would be quite complex. Do you have any ideas for:

1. How I could most properly achieve what I am attempting ("widen" a generic type in a way that the compiler understands) with current TS capabilities? Is the intersection workaround suggested by jcalz my best option?
1. If there's no proper way to do this, is there any reasonable way to enhance TS with the capability to "widen" a generic type in a way that it understands?
